### PR TITLE
fixed regressions with old hiscore plugin

### PR DESCRIPTION
**Pull request by eadmaster**
see https://github.com/mamedev/mame/pull/6130

### CODE REVIEW

File: plugins/hiscore/init.lua [hunk 1]
```diff
@@ -95,17 +95,31 @@ function hiscore.startplugin()
...
+		-- file not found
+		emu.print_verbose( "hiscore file not found: " .. hiscoredata_path );
+		return ""
 	  end
+	  local rm_match;
+	  local rm_match_crc = 0;
 	  if emu.softname() ~= "" then
 		local soft = emu.softname():match("([^:]*)$")
-		rm_match = '^' .. emu.romname() .. ',' .. soft .. ':';
+		rm_match = emu.romname() .. ',' .. soft .. ':';
```
Comment (cuavas) [inline plugins/hiscore/init.lua]: Changing this to not only match the start doesn’t seem like a desirable change.

File: plugins/hiscore/init.lua [hunk 2]
```diff
@@ -158,8 +174,17 @@ function hiscore.startplugin()
 	  local r;
 	  if emu.softname() ~= "" then
 		local soft = emu.softname():match("([^:]*)$")
-		r = hiscore_path .. '/' .. emu.romname() .. "_" .. soft .. ".hi";
+		r = hiscore_path .. '/' .. emu.romname() .. '/' .. soft .. ".hi";
```
Comment (cuavas) [inline plugins/hiscore/init.lua]: Why the change in format here?  This also doesn’t account for the fact that there are three degrees of freedom: machine, list and software.  There are machines with multiple software lists with software items with identical short names in each list.  The `.hi` file format doesn’t seem to account for this either.

File: plugins/hiscore/init.lua [hunk 3]
```diff
@@ -172,6 +197,7 @@ function hiscore.startplugin()
 	  if not output then
 		-- attempt to create the directory, and try again
 		lfs.mkdir( hiscore_path );
+		lfs.mkdir( hiscore_path .. '/' .. emu.romname() );
```
Comment (cuavas) [inline plugins/hiscore/init.lua]: Why are you always creating this path when it’s not going to be used unless stuff is found for the current software?

File: plugins/hiscore/init.lua [hunk 4]
```diff
@@ -283,7 +309,11 @@ function hiscore.startplugin()
 		last_write_time = -10
 		emu.print_verbose("Starting " .. emu.gamename())
 		config_read = read_config();
-		local dat = read_hiscore_dat()
+		local dat = read_hiscore_dat("hiscore.dat")
+		if dat and dat == "" then
+			-- nothing found, try in console_hiscore.dat instead
+			dat = read_hiscore_dat("console_hiscore.dat")
+		end
```
Comment (cuavas) [inline plugins/hiscore/init.lua]: This seems very dubious.  The `hiscore.dat` file will always be present in a normal distribution.  With the code like this, the only way to trigger reading the `console_hiscore.dat` is to remove it.

File: plugins/hiscore/init.lua (hunk 1 above)
Comment (eadmaster) [reply plugins/hiscore/init.lua]: right, i'm fixing this...

File: plugins/hiscore/init.lua (hunk 4 above)
Comment (eadmaster) [reply plugins/hiscore/init.lua]: nope,  `read_hiscore_dat()` will return an empty string if no cluster/data lines are found for the current game. In this case `console_hiscore.dat` is also checked.

File: plugins/hiscore/init.lua (hunk 3 above)
Comment (eadmaster) [reply plugins/hiscore/init.lua]: ok, i'll fix it with this check:
````
		if manager:machine().images["cart"] or manager:machine().images["cdrom"] ~= "" then
			lfs.mkdir( hiscore_path .. '/' .. emu.romname() );
		end
````

File: plugins/hiscore/init.lua (hunk 2 above)
Comment (eadmaster) [reply plugins/hiscore/init.lua]: can you make me an example or point me to some datfiles?

Comment (rb6502) [reply plugins/hiscore/init.lua]: I don't think there are any .hi files for it, but the Apple II machines (hash/apple2_flop*.xml) have multiple software lists with the same shortname, generally "original" (with copy protection intact), "clean cracked" (new modern-day crack with no defacement or removed screens/features), and sometimes just "cracked" (back-in-the-day crack with crack intros etc).

Comment (eadmaster) [reply plugins/hiscore/init.lua]: `flop` devices are not supported currently, btw what `emu.romname()` returns in this case?

Comment (bradyreino) [reply plugins/hiscore/init.lua:178]: @eadmaster Can you patch Johnny Nero: Action Hero MAME by bypassing the unencrypted data so we can re-dump it to be working and playable? (suicide battery means helps the key stored into FPGA on VP101 hardware like the Capcom Power System II hardware that known for Marvel vs. Capcom arcade game). s-l1600

Comment (bradyreino) [reply plugins/hiscore/init.lua:178]: @cuavas Johnny Nero: Action Hero will be playable on MAME 0.224 after patching the unencrypted data on the PCB with hard drive and redumping it.
original (4)
